lib/sparse/src/index/posting_list.rs

```diff
@@ -0,0 +1,178 @@
+use crate::common::types::{DimWeight, RecordId};
+
+#[derive(Debug, Copy, Clone)]
+pub struct PostingElement {
+    pub id: RecordId,
+    pub weight: DimWeight,
+    pub max_next_weight: DimWeight,
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct PostingList {
+    /// List of the posting elements ordered by id
+    pub elements: Vec<PostingElement>,
+}
+
+impl PostingList {
+    #[cfg(test)]
+    pub fn from(records: Vec<(RecordId, DimWeight)>) -> PostingList {
+        let mut posting_list = PostingBuilder::new();
+        for (id, weight) in records {
+            posting_list.add(id, weight);
+        }
+        posting_list.build()
+    }
+}
+
+pub struct PostingBuilder {
+    elements: Vec<PostingElement>,
+}
+
+impl PostingBuilder {
+    pub fn new() -> PostingBuilder {
+        PostingBuilder {
+            elements: Vec::new(),
+        }
+    }
+
+    pub fn add(&mut self, id: RecordId, weight: DimWeight) {
+        self.elements.push(PostingElement {
+            id,
+            weight,
+            max_next_weight: f32::NEG_INFINITY,
+        });
+    }
+
+    pub fn build(mut self) -> PostingList {
+        // Sort by id
+        self.elements.sort_by_key(|e| e.id);
+
+        // Check for duplicates
+        #[cfg(debug_assertions)]
+        {
+            for i in 1..self.elements.len() {
+                if self.elements[i].id == self.elements[i - 1].id {
+                    panic!("Duplicate id {} in posting list", self.elements[i].id);
+                }
+            }
+        }
+
+        // Calculate max_next_weight
+        let mut max_next_weight = f32::NEG_INFINITY;
+        for i in (0..self.elements.len()).rev() {
+            let element = &mut self.elements[i];
+            element.max_next_weight = max_next_weight;
+            max_next_weight = max_next_weight.max(element.weight);
+        }
+
+        PostingList {
+            elements: self.elements,
+        }
+    }
+}
+
+pub struct PostingListIterator<'a> {
+    posting_list: &'a PostingList,
+    current_index: usize,
+}
+
+impl<'a> PostingListIterator<'a> {
+    pub fn new(posting_list: &'a PostingList) -> PostingListIterator<'a> {
+        PostingListIterator {
+            posting_list,
+            current_index: 0,
+        }
+    }
+
+    pub fn peek(&self) -> Option<&PostingElement> {
+        self.posting_list.elements.get(self.current_index)
+    }
+
+    pub fn next(&mut self) -> Option<&PostingElement> {
+        if self.current_index < self.posting_list.elements.len() {
+            let element = &self.posting_list.elements[self.current_index];
+            self.current_index += 1;
+            Some(element)
+        } else {
+            None
+        }
+    }
+
+    pub fn len_left(&self) -> usize {
```

Comment:
I find this name a bit confusing. I find `left` ambiguous here.

Maybe `len_to_end` or `remaining_len` or `remaining_items` is a better name.